### PR TITLE
Add a csv parser based splitter

### PR DESCRIPTION
**Pull request by SamRoberts**

### CODE REVIEW

File: maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala [hunk 1]
```diff
@@ -24,22 +24,69 @@ object SplitterSpec extends test.Spec { def is = s2"""
...
+  val fullyQuotedCsv: Gen[(Int, String)] = for {
+    numSegments <- Gen.choose(0,20)
+    segments    <- Gen.listOfN(numSegments, Gen.oneOf(
+                     Gen.alphaStr,
+                     fullyQuotedField
+                   ))
+  } yield (numSegments, segments.mkString(delimiter))
+
   val delimitedStr: Gen[String] = for {
     numSegments <- Gen.choose(0,20)
```
Comment (stephanh) [inline maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala]: [Codacy](https://www.codacy.com/app/CommBank/maestro/file/448873761/issues/source?bid=2205397&fileBranchId=2231288#l72) Magic Number

File: maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala [hunk 2]
```diff
@@ -47,24 +94,37 @@ Splitter properties
...
 
-  val delimitedRoundTrip: Prop = Prop.forAll(delimitedStr) (str => {
+  val delimitedRoundtrip: Prop = Prop.forAll(delimitedStr) (str => {
     val segments = Splitter.delimited(delimiter).run(str)
     segments.mkString(delimiter) must_== str
   })
 
+  val csvSampleRow = {
+    val sampleRow = """"ABC001","232",45,0.00,"Sample Data","Sample Data (String,Title)","8/298/78899",1,"26/11/2014",1,"","""""
+    val expected = List("ABC001","232","45","0.00","Sample Data","Sample Data (String,Title)","8/298/78899","1","26/11/2014","1","","")
```
Comment (stephanh) [inline maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala]: [Codacy](https://www.codacy.com/app/CommBank/maestro/file/448873761/issues/source?bid=2205397&fileBranchId=2231288#l104) The string literal "1" appears 2 times in the file.

File: project/build.scala [hunk 3]
```diff
@@ -100,7 +100,9 @@ object build extends Build {
           noHadoop("commons-validator"  % "commons-validator" % "1.4.0"),
```
Comment (stephanh) [inline project/build.scala]: [Codacy](https://www.codacy.com/app/CommBank/maestro/file/448873845/issues/source?bid=2205397&fileBranchId=2231288#l100) The string literal "commons-validator" appears 2 times in the file.

File: maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala [hunk 4]
```diff
@@ -24,22 +24,69 @@ object SplitterSpec extends test.Spec { def is = s2"""
...
+                   ))
+  } yield (numSegments, segments.mkString(delimiter))
+
   val delimitedStr: Gen[String] = for {
     numSegments <- Gen.choose(0,20)
     segments    <- Gen.listOfN(numSegments, Gen.alphaStr)
   } yield segments.mkString(delimiter)
 
+  val goodFixedLengthStr: Gen[(List[Int], String)] = for {
+    numFields <- Gen.choose(0, 100)
```
Comment (stephanh) [inline maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala]: [Codacy](https://www.codacy.com/app/CommBank/maestro/file/448873761/issues/source?bid=2205397&fileBranchId=2231288#l77) Magic Number

File: project/build.scala [hunk 5]
```diff
@@ -100,7 +100,9 @@ object build extends Build {
...
           "au.com.cba.omnia"           %% "ebenezer-test"     % ebenezerVersion        % "test",
```
Comment (stephanh) [inline project/build.scala]: [Codacy](https://www.codacy.com/app/CommBank/maestro/file/448873845/issues/source?bid=2205397&fileBranchId=2231288#l101) The string literal "au.com.cba.omnia" appears 2 times in the file.

File: maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala [hunk 6]
```diff
@@ -24,22 +24,69 @@ object SplitterSpec extends test.Spec { def is = s2"""
...
+    lengths   <- Gen.listOfN(numFields, Gen.choose(0, 20))
+    row       <- Gen.listOfN(lengths.sum, Gen.alphaChar).map(_.mkString)
+  } yield (lengths, row)
+
+  val badFixedLengthStr: Gen[(List[Int], String)] = for {
+    numFields <- Gen.choose(0, 100)
+    lengths   <- Gen.listOfN(numFields, Gen.choose(0, 20))
```
Comment (stephanh) [inline maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala]: [Codacy](https://www.codacy.com/app/CommBank/maestro/file/448873761/issues/source?bid=2205397&fileBranchId=2231288#l84) Magic Number

**Comment by SamRoberts** [comment]
In response to codacy's somewhat reasonable issues, I should say that I made some attempt to refactor the generators in this test, but abandoned the attempt, as:

1. My refactored code still contained magic numbers and other things that would cause codacy to complain, and I didn't want to introduce the level of indirection required to completely remove the warnings.
2. Although it is easy to remove duplicated code between the csv and delimited generators, as they are so similar, it is difficult to see how to extend this to the fixed length string generators (especially `badFixedLengthStr`), and I didn't have a lot of enthusiasm for the incomplete solutions I was introducing just to make codacy happier (but still unhappy).

I think if I were going to refactor these, I would want to do a more thorough job, introducing types for the individual fields and csv / fixed length strings being generated, and arbitrary instances for those types, so these tests can tie in with specs2 better, and some notion of "fuzzed" types to create "bad" generators. But it's all a bit too much work for this small suite of tests, IMO.

File: maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala [hunk 7]
```diff
@@ -24,22 +24,69 @@ object SplitterSpec extends test.Spec { def is = s2"""
...
+
+  // no idea what the field count or even character count will be,
+  // but it should be valid csv if the number of quotes is even
+  val goodCsv: Gen[String] =
+    Gen.listOf(Gen.frequency(
+      (8, Gen.alphaNumChar),
+      (1, Gen.const('"')),
+      (1, Gen.const(delimiterChr))
+    ))
+      .suchThat(chrs => chrs.count(_ == '"') % 2 == 0)
```
Comment (stephanh) [inline maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala]: Is that the same as filter?

**Comment by stephanh** [comment]
:+1:

File: maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala (hunk 7 above)
Comment (SamRoberts) [inline maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala]: Exactly. In fact, filter just calls `suchThat`.

Comment (stephanh) [inline maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala]: You don't want to just use filter as the name instead then. That way everyone already knows what to expect and doesn't wonder why it is different?

Comment (SamRoberts) [inline maestro-core/src/test/scala/au/com/cba/omnia/maestro/core/split/SplitterSpec.scala]: Yeah, fair enough. `suchThat` is actually the name I am used to in a quickcheck context, but that's not going to be true for most people.

**Pull request merged**